content/blog/2023-12-11-array-ir/index.md

```diff
@@ -0,0 +1,109 @@
+
+We decided that control flow instructions should be statements rather than expressions. This is because all expressions are associated with a protocol, but control flow must be executed outside of any cryptographic protocol. 
+
+Implementing control flow required a slight change to how “blocks” (sequences of code delimited by curly braces) are represented in our AST. Previously, blocks were only used to represent function bodies, and therefore required that the last statement was always a `return`. We decided to introduce distinct notions of blocks for function bodies and control flow branches. This is because the invariant that each function contains a single unique `return` statement makes code generation far more straightforward, and doesn’t sacrifice expressivity. 
+
+We evaluated this implementation using handwritten tests to ensure all branches through control flow are properly executed.
+
+# Precircuit IR
+We then implemented a “pre-circuit IR”. This is intended to be an updated version of the output of the original Viaduct project, which produces a program where each let-binding is annotated with the protocol that will execute it. This had to be implemented separately since AIRduct represents computations differently from the original Viaduct, being more array-based and having support for array operations.
+
+The pre-circuit IR represents functions as a sequence of control flow statements and let-bindings. Circuits are not present in this IR. 
+
+The pre-circuit IR is different from the circuit IR in that:
+1. All let-bindings are annotated with a protocol. In the circuit IR, only let-bindings involving a “command” (input, output, and function/circuit calls) are annotated with protocols denoting where results are stored. Let-bindings involving computations occur within circuits, which are annotated with the associated protocol for all included statements. 
+2. It includes downgrade nodes. These are explicit violations of secure information flow, such as declassifying secret data or endorsing untrusted values. It is necessary that downgrade instructions are preserved in the precircuit IR, since they enforce a secure ordering on computations.
+
+# Grouping
+The grouping procedure places statements of the same protocol together whenever possible. This will enable larger circuit blocks, reducing the overhead of communication between protocols.
+
+This reordering is subject to data and security dependencies. Node `x` is a dependency of node `y` when:
+- `x` defines data that is used by `y`,
+- `x` is an `input` statement, and `y` an output,
+- `x` is an `endorse` statement, and `y` an declassify, or
+- `x` is an `output` statement, and `y` an input from the same host.
+
+The first rule corresponds to data dependencies. The second and third rules hold because moving a statement which reveals information before one that commits to data violates security. The last rule preserves the user interface.
```

Comment:
Maybe this is silly, but what about an `output` followed by an `output`? Do those have to have their order preserved? (This would be true for memory writes, for example, because you want to preserve "overwrites" of the same memory location. But maybe this is not possible with AIRduct outputs… if so, maybe the blog post needs a little more background on what "inputs" and "outputs" are in this setting?)